File: akasha-master/packages/arcasha-router-py/arcasha_router/bayesian.py
"""Bayesian belief (state estimation) — ArcAsha ODAR core."""

import math


class BayesianBelief:
    """Running-mean Bayesian estimator for per-(node, capability) quality.

    confidence(n) = 1 - exp(-n/8), effective = mu * confidence.
    Optional prior seed (Closed Bayesian Loop: memory -> mu0 -> observation -> mu).
    """

    def __init__(self, mu: float = 0.5, n: int = 0, seed: dict | None = None):
        if seed is not None and seed.get("n", 0) > 0:
            self.mu = max(0.0, min(1.0, float(seed["mu"])))
            self.n = int(seed["n"])
        else:
            self.mu = float(mu)
            self.n = int(n)

    def update(self, x: float) -> None:
        """Bayesian update with observation x in [0, 1] (sample mean)."""
        self.mu = (self.n * self.mu + x) / (self.n + 1)
        self.n += 1

    def confidence(self) -> float:
        return 1.0 - math.exp(-self.n / 8.0)

    def effective(self) -> float:
        return self.mu * self.confidence()

    def snapshot(self) -> dict:
        return {
            "mu": round(self.mu, 3),
            "n": self.n,
            "confidence": round(self.confidence(), 3),
            "effective": round(self.effective(), 3),
        }


class EmaLatency:
    """Exponential moving average latency smoother (alpha=0.3)."""

    def __init__(self, alpha: float = 0.3):
        self.alpha = alpha
        self.ema = 0.0
        self.n = 0

    def observe(self, measured: float) -> int:
        self.ema = measured if self.n == 0 else self.alpha * measured + (1 - self.alpha) * self.ema
        self.n += 1
        return round(self.ema)

    def value(self) -> int:
        return round(self.ema)

File: akasha-master/packages/arcasha-router/packages/belief/src/bayesian.ts
/**
 * arcasha-router — Bayesian Belief (EXP-0002D.1 / 0003 系)
 *
 * 観測から信念を更新し、confidence (1-exp(-n/8)) と effective (μ×confidence) を提供。
 * Belief-Driven AI Orchestration の核心 (状態推定)。
 */

export interface BeliefSnapshot {
  mu: number;
  n: number;
  confidence: number;
  effective: number;
}

export interface BeliefSeed {
  mu: number;
  n: number;
}

export class BayesianBelief {
  private mu = 0.5;
  private n = 0;

  /**
   * 事前分布シード: Long-term Memory から集計した μ₀/n₀ で初期化
   * (Closed Bayesian Loop: Memory → Prior μ₀ → Observation → Posterior μ)
   */
  constructor(seed?: BeliefSeed) {
    if (seed && seed.n > 0) {
      this.mu = Math.max(0, Math.min(1, seed.mu));
      this.n = Math.floor(seed.n);
    }
  }

  /** 観測 x (0-1) でベイズ更新 (標本平均) */
  update(x: number): void {
    this.mu = (this.n * this.mu + x) / (this.n + 1);
    this.n += 1;
  }

  /** 信頼度: 観測数に応じて 0 → 1 */
  confidence(): number {
    return 1 - Math.exp(-this.n / 8);
  }

  /** 効果値: μ × confidence (二段階の信頼度加重) */
  effective(): number {
    return this.mu * this.confidence();
  }

  snapshot(): BeliefSnapshot {
    const mu = Math.round(this.mu * 1000) / 1000;
    const confidence = Math.round(this.confidence() * 1000) / 1000;
    const effective = Math.round(this.effective() * 1000) / 1000;
    return { mu, n: this.n, confidence, effective };
  }
}

/** EMA によるレイテンシ平滑化 */
export class EmaLatency {
  private ema = 0;
  private n = 0;

  observe(measured: number): number {
    const alpha = 0.3;
    this.ema = this.n === 0 ? measured : alpha * measured + (1 - alpha) * this.ema;
    this.n += 1;
    return Math.round(this.ema);
  }

  value(): number {
    return Math.round(this.ema);
  }
}

